### ISSUE TITLE
Item index exceeds when changing `total` prop

### ISSUE THREAD

**Issue by Artoria-0x04** [petyosi/react-virtuoso#126]
I discovered some rendering inconsistency when changing total count.

One is items are lost when both changing `total` and `item list` to a larger value, total items in the list does not match the count after change, which I haven't reproduced with examples.

Another is the `list` still using last index range when `total` change to a smaller value, which will leads to undefined items. An example below, click the toggle at the bottom of the list.

[https://stackblitz.com/edit/react-ter3me](https://stackblitz.com/edit/react-ter3me)

**Comment by petyosi** [comment]
Seems like a bug, filing as one.

**Comment by petyosi** [comment]
@Artoria-0x04 I played around with the example. The incorrect behavior is related to the fact that the internal **10 list items are re-rendered first when the DataContext value changes** - the outer component still works with the totalCount being 10. I hope that this makes sense. If it does not, the workaround I may offer [is this one](https://stackblitz.com/edit/react-ter3me-vpszeb?file=example.js). Notice how I avoid the context usage by using `useCallback`.

Given that this is not the first time someone struggles with accessing data in the list item renderer, I am thinking about adding data context helper, similar to react-virtualized. 

Please let me know if the workaround works for you. I would like to edit the issue if so, so that people with similar problems can find it.

**Comment by Artoria-0x04** [comment]
@petyosi Thanks for the quick response, it does the trick. I guess a different context item renderer is a must avoid then. Besides, I do think that a vlist api doesn't accept `data list` and an item renderer doesn't provide the `item` is unintuitive, sort of.

**Comment by petyosi** [comment]
@Artoria-0x04 agreed; keeping the data away from the component has its benefits abstraction wise but also causes troubles and makes the component harder to use sometimes.
